### PR TITLE
Add elastic utilities

### PR DESCRIPTION
**Pull request by MFraters**
This pull request splits the tensor functions out of #4987 and #5116 and puts it in the utilities function in a new Tensors namespace.

### CODE REVIEW

File: include/aspect/utilities.h [hunk 1]
```diff
@@ -1217,6 +1217,60 @@ namespace aspect
...
+     * Contains utility functions related to tensors.
+     */
+    namespace Tensors
+    {
+
+      /**
+       * Rotate a 3D 4th order tensor with an other 3D 2th order tensor
+       */
+      SymmetricTensor<4,3>
+      rotate_4th_order_tensor(const SymmetricTensor<4,3> &input_tensor, const Tensor<2,3> &rotation_tensor);
```
Comment (bangerth) [inline include/aspect/utilities.h:1231]: I'm reading this as "apply this rotation to that object", which would suggest to reverse the order of arguments.

File: include/aspect/utilities.h [hunk 2]
```diff
@@ -1217,6 +1217,60 @@ namespace aspect
...
+
+
+      /**
+       * Rotate a 6x6 voigt matrix with an other 3D 4th
```
Comment (bangerth) [inline include/aspect/utilities.h:1235]: incomplet

File: include/aspect/utilities.h [hunk 3]
```diff
@@ -1217,6 +1217,60 @@ namespace aspect
...
+       */
+      SymmetricTensor<2,6>
+      rotate_6x6_matrix(const SymmetricTensor<2,6> &input_tensor, const Tensor<2,3> &rotation_tensor);
```
Comment (bangerth) [inline include/aspect/utilities.h:1238]: same here

File: include/aspect/utilities.h [hunk 4]
```diff
@@ -1217,6 +1217,60 @@ namespace aspect
...
+
+      /**
+       * Transform a 4th order tensor into a 6x6 matrix
```
Comment (bangerth) [inline include/aspect/utilities.h:1241]: Period at the end of the sentence for all of the doc strings of all of these functions.

File: source/utilities.cc [hunk 5]
```diff
@@ -3097,6 +3097,281 @@ namespace aspect
...
+                      {
+                        for (unsigned short int j1 = 0; j1 < 3; j1++)
+                          {
+                            for (unsigned short int j2 = 0; j2 < 3; j2++)
+                              {
+                                for (unsigned short int j3 = 0; j3 < 3; j3++)
+                                  {
+                                    for (unsigned short int j4 = 0; j4 < 3; j4++)
+                                      {
+                                        output[i1][i2][i3][i4] = output[i1][i2][i3][i4] + rotation_tensor[i1][j1]*rotation_tensor[i2][j2]*rotation_tensor[i3][j3]*rotation_tensor[i4][j4]*input_tensor[j1][j2][j3][j4];
```
Comment (bangerth) [inline source/utilities.cc:3126]:
```suggestion
                                        output[i1][i2][i3][i4] += rotation_tensor[i1][j1]*rotation_tensor[i2][j2]*rotation_tensor[i3][j3]*rotation_tensor[i4][j4]*input_tensor[j1][j2][j3][j4];
```

File: source/utilities.cc [hunk 6]
```diff
@@ -3097,6 +3097,281 @@ namespace aspect
...
+      }
+
+
+
+      SymmetricTensor<2,6>
+      rotate_6x6_matrix(const SymmetricTensor<2,6> &input_tensor, const Tensor<2,3> &rotation_tensor)
+      {
+        // we can represent the rotation of the 4th order tensor as a rotation in the Voigt
+        // notation by computing $C'=MCM^{-1}$. Because M is orthogonal we can replace $M^{-1}$
+        // with $M^T$ resutling in $C'=MCM^{T}$ (Carcione, J. M. (2007). Wave Fields in Real Media:
```
Comment (bangerth) [inline source/utilities.cc:3146]:
```suggestion
        // with $M^T$ resulting in $C'=MCM^{T}$ (Carcione, J. M. (2007). Wave Fields in Real Media:
```

**Review by bangerth — COMMENTED** [review]
Very nice! In particular, I kept thinking that we ought to have tests for this, and then what do you know, there are tests at the bottom of the patch!

There are different ways of representing 3x3x3x3 tensors as 6x6 matrices, and the same applies to 3x3 tensors as vector of size 6. Voigt notation is one. Take a look at https://en.wikipedia.org/wiki/Voigt_notation.

I think it would be useful to annotate each function how exactly you want that representation to act -- perhaps by linking to the Wikipedia article.

**Review by gassmoeller — COMMENTED** [review]
A few more comments.

**Comment by bangerth** [comment]
As mentioned earlier today, there seems to be a trend for functions of the form `to_X`, which omits the word `convert_` and gets the type of the object being converted from the argument. An example is `std::to_string`.

**Comment by MFraters** [comment]
I have changed the names

**Review by gassmoeller — COMMENTED** [review]
I dont mind the name change, but the new names do not seem consistent to me.

**Comment by MFraters** [comment]
sorry about that, I fixed it.

**Review by gassmoeller — APPROVED** [review]

**Review by bangerth — APPROVED** [review]

**Pull request merged**